client_credentials_auth.go

```diff
@@ -0,0 +1,129 @@
+package sdk
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+)
+
+type ClientCredentialsTokenSource struct {
+	clientID     string
+	clientSecret string
+	tokenURL     string
+	scope        string
+	grantType    string
+	token        *ClientCredentialsToken
+	lock         sync.RWMutex
+}
+
+type ClientCredentialsAuth struct {
+	tokenSource *ClientCredentialsTokenSource
+}
+
+func (cca *ClientCredentialsAuth) Set(req *http.Request) error {
+	token, err := cca.tokenSource.Token()
+	if err != nil {
+		return err
+	}
+	req.Header.Add("Authorization", "Bearer "+token)
+	return nil
+}
+
+func NewClientCredentialsTokenSource(clientID, clientSecret, tokenURL, scope, grantType string) *ClientCredentialsTokenSource {
+	return &ClientCredentialsTokenSource{
+		clientID:     clientID,
+		clientSecret: clientSecret,
+		tokenURL:     tokenURL,
+		scope:        scope,
+		grantType:    grantType,
+	}
+}
+
+func (ts *ClientCredentialsTokenSource) Token() (string, error) {
+	ts.lock.RLock()
+	expired := ts.token == nil || ts.token.Expired()
+
+	if expired {
+		ts.lock.RUnlock()
+
+		token, err := obtainClientCredentialsToken(ts.clientID, ts.clientSecret, ts.tokenURL, ts.scope, ts.grantType)
+		if err != nil {
+			return "", err
+		}
+
+		ts.lock.Lock()
+		ts.token = token
+		ts.lock.Unlock()
+
+		return token.AccessToken, nil
+	}
+
+	ts.lock.RUnlock()
+	return ts.token.AccessToken, nil
+}
+
+func obtainClientCredentialsToken(clientID, clientSecret, tokenURL, scope, grantType string) (*ClientCredentialsToken, error) {
+
+	reqBody := url.Values{}
+	reqBody.Set("client_id", clientID)
+	reqBody.Set("client_secret", clientSecret)
+	reqBody.Set("grant_type", grantType)
+	reqBody.Set("scope", scope)
+
+	buffer := []byte(reqBody.Encode())
+
+	req, err := http.NewRequest(http.MethodPost, tokenURL, bytes.NewBuffer(buffer))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	var body []byte
+	if res.Body != nil {
+		defer res.Body.Close()
+		body, _ = io.ReadAll(res.Body)
+	}
+
+	if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code %d, body: %s", res.StatusCode, string(body))
+	}
+
+	token := &ClientCredentialsToken{}
```

Comment:
I think Ideally we could also reuse the token struct that already has the `Expired()` method instead of duplicating this code.

After all the Token does represent a default OAuth token format.

```go
    tj := &tokenJSON{}
	if err := json.Unmarshal(body, tj); err != nil {
		return nil, fmt.Errorf("unable to unmarshal token: %s", err)
	}

	return &Token{
		IDToken: tj.AccessToken,
		Expiry:  tj.expiry(),
	}
```

The `ClientCredentials` struct can than be removed.

Reply:
That sounds like a good idea potentially, however it's not an ID token that we're getting back from the client credentials flow?

Reply:
Something we can try instead is to use the TokenSource interface within the new code, so they share a common way of working.